### PR TITLE
fixes for main

### PR DESCRIPTION
**Pull request by leafyoung**
A few small fixes:
- When dump json with model run, use indent=4 to print multi-line results for reading
- When throw out error, add block_number information. Good for debugging Contract
- Add a local in-memory cache for contract.meta. When creating the same token for different swap protocols during querying token.prices, it's slow and hitting DB each time if without cache,
- Optimize Token._load(), move each property (i.e. symbol, decimals, etc) retrieval code to each property. It may take up to 1.x sec from usual a few micro-seconds. When doing historical query, this could slow things down.
- Add alternative web3 event filtering for some node server.
- Fix for loading instance()/functions, they only return non-None. 
- To allow fungible token query with address, to solve the current token loading issues, mostly for upgradeable tokens.
- Various fixes for linting and pylance pdate Address class 
- Passed tests against all current models.

### CODE REVIEW

File: credmark/credmark_dev.py [hunk 1]
```diff
@@ -433,7 +433,7 @@ def run_model(args):  # pylint: disable=too-many-statements,too-many-branches,to
             else:
                 exit_code = 1
 
-        json_dump(result, sys.stdout)
+        json_dump(result, sys.stdout, indent=4)
```
Comment (iwyrkore) [inline credmark/credmark_dev.py:436]: Please a flag for this and do not do it by default. `--format_json`

File: credmark/model/describe.py [hunk 2]
```diff
@@ -161,18 +161,14 @@ def is_parent(child, parent):
         else:
             validate_model_slug(slug)
 
-        model_desc = description
```
Comment (iwyrkore) [inline credmark/model/describe.py:164]: Why did you remove this?

File: credmark/types/data/contract.py [hunk 3]
```diff
@@ -74,47 +94,113 @@ def _load(self):
...
     @property
     def proxy_for(self):
         if not self._loaded:
             self._load()
-        if self._proxy_for is None and self.is_transparent_proxy:
+
+        if self._proxy_for is None and self.is_transparent_proxy and self.instance is not None:
             context = credmark.model.ModelContext.current_context()
 
+            default_proxy_address = ''.join(['0'] * 40)
```
Comment (newtectonics) [inline credmark/types/data/contract.py:123]: this is cool, but peter was able to get it into the database

File: credmark/credmark_dev.py (hunk 1 above)
Comment (leafyoung) [reply credmark/credmark_dev.py:436]: Sure. I have made it an option.

File: credmark/dto/encoder.py [hunk 4]
```diff
@@ -17,9 +17,9 @@ def default(self, o):
         return json.JSONEncoder.default(self, o)
 
 
-def json_dump(obj, fp):  # pylint: disable=invalid-name
+def json_dump(obj, fp, post_proc=lambda x: x, **b):  # pylint: disable=invalid-name
     """Dump an obj that may contain embedded DTOs to json"""
-    return json.dump(obj, fp, cls=PydanticJSONEncoder)
+    return post_proc(json.dump(obj, fp, cls=PydanticJSONEncoder, **b))
```
[the remark this answers is not in the document]
Comment (leafyoung) [reply credmark/dto/encoder.py]: Sure. I have changed the parameter name to `json_dump_args`. I also moved the string processing function to `json_dumps`.

```
def json_dump(obj, handle, **json_dump_args):
    """Dump an obj that may contain embedded DTOs to json"""
    return json.dump(obj, handle, cls=PydanticJSONEncoder, **json_dump_args)

def json_dumps(obj, format_json=lambda x: x, **json_dump_args):
    """Dump an obj that may contain embedded DTOs to json string"""
    return format_json(json.dumps(obj, cls=PydanticJSONEncoder, **json_dump_args))
```

File: credmark/model/describe.py (hunk 2 above)
Comment (leafyoung) [reply credmark/model/describe.py:164]: The merging was not done properly. I have restored.

File: credmark/types/data/contract.py [hunk 5]
```diff
@@ -13,6 +17,18 @@
 from credmark.types.data.data_content.transparent_proxy_data import UPGRADEABLE_CONTRACT_ABI
 
 
+class Singleton:
```
[the remark this answers is not in the document]
Comment (leafyoung) [reply credmark/types/data/contract.py:22]: Thanks for the ideas! I have added the chain id and a check for block number. If the cache has the meta from a later block, it will not return the cache.

The cache traces are

Case 1) token.price for one block
```
[Cache] Not found in cache chain_id=1/0x1f98431c8ad98523631ae4a59f267346ea31f984/14234904
[Cache] Not found in cache chain_id=1/0x59e1f901b5c33ff6fae15b61684ebf17cca7b9b3/14234904
[Cache] Not found in cache chain_id=1/0xf7a716e2df2bde4d0ba7656c131b06b1af68513c/14234904
[Cache] Return from cache chain_id=1/0x59e1f901b5c33ff6fae15b61684ebf17cca7b9b3/14234904
[Cache] Return from cache chain_id=1/0xf7a716e2df2bde4d0ba7656c131b06b1af68513c/14234904
```

Case 2) Historical token.price
```
[Cache] Not found in cache chain_id=1/0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc/14233425 <= we start from earlier blocks.
[Cache] Not found in cache chain_id=1/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/14233425
[Cache] Return from cache chain_id=1/0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc/14233425
[Cache] Return from cache chain_id=1/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/14233425
...
[Cache] Return from cache chain_id=1/0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc/14233634 <= returns from cache for later blocks.
[Cache] Return from cache chain_id=1/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/14233634
...
```

File: credmark/dto/encoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (leafyoung) [reply credmark/dto/encoder.py]: Because json.dump multi-line string as quoted. The trace log looks like below. Replacement from quoted `\\n` to `\n` is needed.

Original
```
"trace": "  File \"/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/contract.py\", line 245, in proxy_for\n    raise ModelDataError(f'Unable to retrieve abi for proxy for {self.address}')\n  File \"/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/contract.py\", line 141, in _load\n    if self.proxy_for is not None:\n  File \"/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/token.py\", line 98, in _load\n    super()._load()\n  File \"/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/contract.py\", line 167, in proxy_for\n    self._load()\n  File \"/home/yangye/dev/credmark/credmark-models-py/./models/credmark/tokens/token.py\", line 91, in run\n    _ = input.proxy_for\n"
```

With the string process lambda:
```
 File "/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/contract.py", line 245, in proxy_for
    raise ModelDataError(f'Unable to retrieve abi for proxy for {self.address}')
  File "/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/contract.py", line 141, in _load
    if self.proxy_for is not None:
  File "/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/token.py", line 98, in _load
    super()._load()
  File "/home/yangye/dev/credmark/credmark-models-py/../credmark-model-framework-py/credmark/types/data/contract.py", line 167, in proxy_for
    self._load()
  File "/home/yangye/dev/credmark/credmark-models-py/./models/credmark/tokens/token.py", line 91, in run
    _ = input.proxy_for
```

File: credmark/types/data/contract.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (leafyoung) [reply credmark/types/data/contract.py:22]: @newtectonics, the contract.meta may need an update. It does not use the block_number now.

```
  async contractMetadataLookup(
    connection: Snowflake,
    chainId: number,
    _blockNumber: number,
    input: ContractMetadataQueryInput,
  ): Promise<ContractMetadataQueryOutput> {
    // IMPORTANT: If you update the logic to handle more parameters,
    // update the docs in the register call above.

    if (input.contractAddress) {
      return {
        contracts: await getContractMetaByAddress(
          connection,
          chainId,
          input.contractAddress,
        ),
      };
    }
...
  }
```

**Comment by iwyrkore** [comment]
@leafyoung You don't need to update this PR. We'll copy over some of the features into separate PRs.

File: credmark/dto/encoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (iwyrkore) [reply credmark/dto/encoder.py]: But whatever is calling `json_dumps()` can run whatever processing they need on the output. 

`format_json_whatever(json_dumps(obj))`

[the remark this answers is not in the document]
Comment (leafyoung) [reply credmark/dto/encoder.py]: ok. I have remove the lambda.

**Comment by iwyrkore** [comment]
json formatting and error message detail for exceptions during model run are added in framework version 0.8.3.

**Comment by leafyoung** [comment]
Close this PR as parts of it has or will be merged to main.
